kernel/idle.c

```diff
@@ -157,7 +157,17 @@ void idle(void *unused1, void *unused2, void *unused3)
 		k_yield();
 #else
 		(void)arch_irq_lock();
-		sys_power_save_idle();
+
+		/*
+		 * Check if another thread is ready due to interrupts
+		 * taken before they were locked.
+		 */
+		if (_kernel.ready_q.cache == _current) {
```

Comment:
This isn't correct on SMP.  There's no one "cache" variable there, every CPU needs to check the queue dynamically to see if it should be preempted.

If you want to do this correctly, you want to duplicate the logic in k_yield() and then call sys_power_save_idle() only if the yield didn't swap. I guess it would have to take a lock key along the lines of reschedule() and release it.

Reply:
> This isn't correct on SMP. There's no one "cache" variable there, every CPU needs to check the queue dynamically to see if it should be preempted.
> 
> If you want to do this correctly, you want to duplicate the logic in k_yield() and then call sys_power_save_idle() only if the yield didn't swap. I guess it would have to take a lock key along the lines of reschedule() and release it.

Agreed.  I did not consider SMP . This is just conceptual patch to confirm the race.

Reply:
Do you think 
` if (z_get_next_ready_thread() == _current)` 
will take care of SMP as well.  But I am not sure it probably does more than needed and what side effect it can have checking it here but not doing a context switch.

Reply:
@andyross  any thoughts on this api usage like this.
If we are not going this way then forcing idle thread to be preempt-able is the option. That will enable 
`CONFIG_PREEMPT ` . Can that have any side effect for a system designed to run on co-operative only mode.

Reply:
Most of the platforms no longer care about the distinction.  ARM still does, there are something like 6 instructions on the interrupt exit path required to choose the next thread that aren't required when !PREEMPT.  But again, in modern usage this is an increasingly obscure configuration.

If we want those cycles back, let's find them somewhere else via optimization.